Upgrade GlacierScript to python3

bitcoinhodler opened the pull request:
GlacierScript was unnecessarily using an old version of Python (2.7.12), while Python 3.5.2 is available as `python3`.

I successfully ran all the developer tests on my own live boot of Ubuntu 16.04.1 (same as the protocol specifies).

Let's bring GlacierScript into modern times with Python 3.

As usual, I recommend that reviewers go one commit at a time, as the commit messages explain both what & why for each change.

bitcoinhodler commented:
For some reason, probably having to do with all the rebases I did on this branch before creating this PR, GitHub is listing the commits out of order. The correct order is:

1a806ba Revert "Swapping base 58 implementation for Gavin Andresen’s"
62def07 Update base58.py to latest released version
1181e75 Update link to base58.py
ea3f95c Add copyright notice to base58.py
60e189b Use OrderedDict for addresses dict, for stability
7904667 Use OrderedDict for stability
ac421ec Use modern, python3-ish print statements
dff8ba6 Switch to python3
6a01c4f Update for python3
5b5d821 Use modern python3 module
bb0a435 Alphabetize module imports, as recommended by PEP 8
d335153 Remove unneeded module import

Opiumreg18 commented:
> По какой-то причине, вероятно, имея дело со всеми переизданиями, которые я сделал на этой ветке, прежде чем создать этот PR, GitHub перечисляет коммиты из строя. Правильный порядок:
> 
> [1a806ba](https://github.com/GlacierProtocol/GlacierProtocol/commit/1a806ba5c4fb104f8e8ba5c9b43d4d621d7ec477) вернуться "замена базы 58 реализации Гэвин Андерсен"
> [62def07](https://github.com/GlacierProtocol/GlacierProtocol/commit/62def076239f0b3a616cf84067f61b13d108fd1d) обновление base58.py в последней версии
> [1181e75](https://github.com/GlacierProtocol/GlacierProtocol/commit/1181e756f5a99bc65ab5d19ce07b11bb0b2c7199) обновить ссылку base58.py
> [ea3f95c](https://github.com/GlacierProtocol/GlacierProtocol/commit/ea3f95ccf501bbb12d6ec5fd03cf1432c472d657) добавить уведомление об авторском праве на base58.py
> [60e189b](https://github.com/GlacierProtocol/GlacierProtocol/commit/60e189b4883cbe6ea43e2304a02157372d53c764) использовать OrderedDict для адресов дикт, для стабильности
> [7904667](https://github.com/GlacierProtocol/GlacierProtocol/commit/79046674beac6e4282962cc136335c072b6e63cb) использовать OrderedDict для стабильности
> [ac421ec](https://github.com/GlacierProtocol/GlacierProtocol/commit/ac421ec17d8d8e0e5b63ac1e7584e5b87bd56c69) использовать современные питон3-ишь печати заявления
> [dff8ba6](https://github.com/GlacierProtocol/GlacierProtocol/commit/dff8ba63bb5afe13270ef447076ead767ee1e6cc) переключиться на питон3
> [6a01c4f](https://github.com/GlacierProtocol/GlacierProtocol/commit/6a01c4f893635b9c1bc2baaf276f4173acff19c5) обновление для питон3
> [5b5d821](https://github.com/GlacierProtocol/GlacierProtocol/commit/5b5d8214cea04f6c1768082da2e735ea49f6d021) использовать современные питон3 модуль
> [bb0a435](https://github.com/GlacierProtocol/GlacierProtocol/commit/bb0a4351d9c4e3e48fed5f0589c98e7bb0d2f797) алфавиту модуль импорта, как рекомендуют ПЭП 8
> [d335153](https://github.com/GlacierProtocol/GlacierProtocol/commit/d3351535346be3ae754400d9d5ccff98e63ccc0b) удалить ненужные модуль импорта

gracenoah commented on base58.py:
```diff
@@ -23,108 +24,168 @@
+alphabet = b'123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz'
+
+
+if bytes == str:  # python2
```
can we use the sys package instead to check the python version?  https://stackoverflow.com/questions/9079036/how-do-i-detect-the-python-version-at-runtime

edit: if this is copied from elsewhere we should have a clear link to it, preferably in the file itself.

bitcoinhodler replied:
The source of this code is explained in the commit message for 1a806ba5c4fb104f8e8ba5c9b43d4d621d7ec477, and 62def076239f0b3a616cf84067f61b13d108fd1d, and in commit 1181e756f5a99bc65ab5d19ce07b11bb0b2c7199 the link was added to the glacierscript.py source.

gracenoah commented on base58.py:
```diff
@@ -1,20 +1,21 @@
+with the bitcoin network.
+'''
 
+# Copyright (c) 2015 David Keijser
```
Is this now a combination of the old script and https://github.com/trezor/trezor-core/blob/master/src/trezor/crypto/base58.py ?

gracenoah commented on t/create-withdrawal-data.2-of-3-nonsegwit.golden:
```diff
@@ -36,8 +36,8 @@ Is this data correct?
 0.20000000 BTC in unspent supplied transactions
-0.19996310 BTC going to destination address mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99
 0 BTC going back to cold storage address 2NGPJX8kzdRpAQJuZWMpCBth1umjQFHeFcz
+0.19996310 BTC going to destination address mxBQD1QAYpwiudaCJdRhE9QSW9cokafJ99
```
I don't understand why this should change with a python version bump?

bitcoinhodler replied:
This is the same issue from 60e189b4883cbe6ea43e2304a02157372d53c764, the ordering of the outputs changed for some of the withdrawals. It was effectively random before; now it is consistent and predictable.

gracenoah commented on t/create-withdrawal-data.uncompressed.decoded:
```diff
@@ -1,6 +1,6 @@
-  "txid": "cc0871827f83c927d79ca8d50c52a72fcaf4223f1de7f92931bfd70f7d44e3b3",
-  "hash": "cc0871827f83c927d79ca8d50c52a72fcaf4223f1de7f92931bfd70f7d44e3b3",
+  "txid": "31906179a4f3d983f62cecd9e8fcd4bc3df4969e3d5ec85547b3fecbdb53c51c",
+  "hash": "31906179a4f3d983f62cecd9e8fcd4bc3df4969e3d5ec85547b3fecbdb53c51c",
```
why did all of these fixtures change?  Should they not be compatible with the script after the version bump?

bitcoinhodler replied:
This is explained in the commit message for 60e189b4883cbe6ea43e2304a02157372d53c764. The order of the transaction outputs was dependent on the ordering of a python dict, which is not specified by the language, and effectively random. It may change when the python interpreter version changes. To alleviate this, I switched to using an OrderedDict, and now the ordering of the outputs is consistent.

gracenoah replied:
got it.  How did we not have tests fail from the unordered python dict before?  Please let me know if I'm missing something

bitcoinhodler replied:
[More than you ever wanted to know about Python dict ordering](https://stackoverflow.com/a/15479974)

In 2.7 it was stable but unspecified. Every run would be the same. The Python3 version we have (3.5.2) uses the random hash seed explained in the SO answer linked above. Every time we run the script, the ordering is randomized, which makes it impossible for us. We need a stable ordering for both the developer tests and so the two quarantined laptops will match each other. Hence OrderedDict.

In Python 3.7 the dict ordering is specified, and we wouldn't need OrderedDict anymore.

bitcoinhodler commented:
Thank you for the review @gracenoah. While I believe all your questions were already answered in the individual commit messages, I've also addressed your comments above.

jazarija commented:
Following are some comments about the pull request

commit: 62def076239f0b3a616cf84067f61b13d108fd1d

- (l. 20-31) If our aim is to switch to python3 then why keep the if-then-else check for python2 ?

Generally speaking I am a bit skeptical of changing the base58 encoding file to this new one. Are we sure the sourced file is coming from a reliable place? Note also how the source file has a cluttered main function that accepts command line arguments etc... Looks like we need to discuss this more.

commit dff8ba63bb5afe13270ef447076ead767ee1e6cc

- l407 Why is the call to list() required here?

commit bb0a4351d9c4e3e48fed5f0589c98e7bb0d2f797

Do we have a convention for when to use "from X import Y" or "import X" and then referencing with either Y or X.Y? Seems like we are confusing two styles with no consistent basis

bitcoinhodler commented:
You can make comments directly on the source code lines during a review, like @gracenoah did above. That would make it easier for me to reply to each individually.

Re 62def07, the old base58.py did not support python3, so I updated it to a later version from a "reliable" open-source repo. I didn't want to make any edits to that file, only copy it verbatim. It has some unneeded stuff, yes, but I figured it was better to use it as-is than to make edits which would require further scrutiny.

Re dff8ba6, the python 2to3 conversion program made this change, and [it's important](https://stackoverflow.com/questions/17695456/why-does-python-3-need-dict-items-to-be-wrapped-with-list) since the list modifies the dict inside the loop. If I remove it, python complains:
```
RuntimeError: OrderedDict mutated during iteration
```

Re bb0a435: I'm not aware of any style guidelines for imports like this (though I haven't gone looking, but pylint et al have no checks for this). Nor do I think any consistency is needed, since it's clear what's being imported in each case. (But `from foo import *` should be discouraged.) I tend to import the entire module if I'm going to use several things from that module, or import just one thing if that's all I need.

Thank you for the detailed review.

jazarija commented:
I see. Regarding the list remark a more direct way would perhaps be to change the for loop and if statement to

`destinations = OrderedDict({key:val for key in destinations if destinations[key] != 0})`

As far as the reliability of the base58 code, how was that determined? I am a bit skeptical to just include code from another repo unless we have an objective measure of its reliability. 

Otherwise, I'd much rather see that we take the code, refactor it, include it as our internal library, remove whatever is not needed and actually properly review what was created.

bitcoinhodler commented:
> `destinations = OrderedDict({key:val for key in destinations if destinations[key] != 0})`

I tried the following:
```
 destinations = OrderedDict({key:val for key, val in destinations.items() if val != '0'})
```

The problem is, this creates a new dict, then passes that dict into OrderedDict. In python 3.5.2, dict ordering is random, which creates problems. (See discussion above.) That's why I switched to OrderedDict in the first place.

For `base58.py`, the repo I sourced this from is the same repo the original Glacier developers sourced their original `base58.py` file from. Later it was changed to a different, older version; unfortunately the git commit history does not record any rationale for this change. The [repo](https://github.com/keis/base58) is a widely used python base58 implementation. What kind of objective measure of reliability is there besides review?

But I take your point, so I pushed some changes to a [new branch](https://github.com/bitcoinhodler/GlacierProtocol/tree/python3-base58-clean) in my fork. Do you think I should include those changes here? The file is greatly simplified now.

jazarija commented:
Gotcha, missed the ordering part. In that case I think you can do

` destinations = OrderedDict( (key, val) for key, val in destinations.items() if val != '0')`

to achieve the required ordering. 

If by including them here you mean in this PR, then I'd say yes. Personally I like this structuring much better this way. We can then take a more in depth look into the code and see if all fits.

bitcoinhodler commented:
Done. @gracenoah can you please renew your review, considering my latest commits?

jlopp commented:
I get a failure while trying to run tests in this branch:

```
GlacierProtocol-python3$ make
cd testrun/create-withdrawal-data.2-of-3-nonsegwit.test && ../../t/create-withdrawal-data.2-of-3-nonsegwit.run 18334   2>&1 > ../../t/create-withdrawal-data.2-of-3-nonsegwit.out
cd testrun/create-withdrawal-data.2-of-3-segwit.test && ../../t/create-withdrawal-data.2-of-3-segwit.run 18335   2>&1 > ../../t/create-withdrawal-data.2-of-3-segwit.out
cd testrun/create-deposit-data.test && ../../t/create-deposit-data.run 18333  2>&1 > ../../t/create-deposit-data.out
cd testrun/create-withdrawal-data.3-of-5-nonsegwit.test && ../../t/create-withdrawal-data.3-of-5-nonsegwit.run 18336   2>&1 > ../../t/create-withdrawal-data.3-of-5-nonsegwit.out
cd testrun/create-withdrawal-data.3-of-5-segwit.test && ../../t/create-withdrawal-data.3-of-5-segwit.run 18337   2>&1 > ../../t/create-withdrawal-data.3-of-5-segwit.out
cd testrun/create-withdrawal-data.bech32.test && ../../t/create-withdrawal-data.bech32.run 18338   2>&1 > ../../t/create-withdrawal-data.bech32.out
cd testrun/create-withdrawal-data.compressed.test && ../../t/create-withdrawal-data.compressed.run 18339   2>&1 > ../../t/create-withdrawal-data.compressed.out
cd testrun/create-withdrawal-data.fails.test && ../../t/create-withdrawal-data.fails.run 18340   2>&1 > ../../t/create-withdrawal-data.fails.out
Line 25 of t/create-withdrawal-data.fails.golden.re failed to match t/create-withdrawal-data.fails.out
Test t/create-withdrawal-data.fails.test failed
make: *** [Makefile:78: t/create-withdrawal-data.fails.test] Error 1
make: *** Waiting for unfinished jobs....

```

bitcoinhodler commented:
Can you paste the last line of `t/create-withdrawal-data.fails.out`? That seems to be the line that is mismatching and it is expected to look like so:

```
subprocess.CalledProcessError: Command '['bitcoin-cli', '-testnet', '-rpcport=18340', '-datadir=bitcoin-test-data', 'createrawtransaction', '[{"txid": "e0e9bb25fb873c4caccdc8ab743c4350310031f2cc077bb90c3f495458860157", "vout": 1}]', '{"2N93du8YobdgsHyu3qgBvSyhGUT52utMNeA": 0, "myP4xdJNwAW9iMakvCjnozg814ewgn8apx": 0}']' returned non-zero exit status 5
```

bitcoinhodler commented:
Also I noticed that in case of a failure like this one, the reported line number was incorrect (off by one), so I just pushed a change to fix that.

jlopp commented:
```
Enter fee rate.
Satoshis per vbyte: Traceback (most recent call last):
  File "../../glacierscript.py", line 864, in <module>
    withdraw_interactive()
  File "../../glacierscript.py", line 747, in withdraw_interactive
    source_address, keys, addresses, redeem_script, txs)
  File "../../glacierscript.py", line 479, in get_fee_interactive
    source_address, destinations, redeem_script, input_txs)
  File "../../glacierscript.py", line 415, in create_unsigned_transaction
    json.dumps(destinations)).strip()
  File "../../glacierscript.py", line 129, in bitcoin_cli_checkoutput
    if retcode != 0: raise subprocess.CalledProcessError(retcode, cmd_list, output=output)
subprocess.CalledProcessError: Command '['bitcoin-cli', '-testnet', '-rpcport=18340', '-datadir=bitcoin-test-data', 'createrawtransaction', '[{"txid": "e0e9bb25fb873c4caccdc8ab743c4350310031f2cc077bb90c3f495458860157", "vout": 1}]', '{"2N93du8YobdgsHyu3qgBvSyhGUT52utMNeA": 0, "myP4xdJNwAW9iMakvCjnozg814ewgn8apx": 0}']' returned non-zero exit status 5.
```

bitcoinhodler commented:
Looks like the difference is in the very last byte, a period at the end of the error message from python. The quarantined laptops with Ubuntu 16.04.1 have python 3.5.2 so I've been testing with 3.5.3 which I already had installed locally. I tried 3.7.4 and I saw the same failure.

I pushed a change to optionally allow the period at the end of the line and now the tests pass for me with both 3.5.3 and 3.7.4.

bitcoinhodler commented:
Also, I have a change in another branch that improves this cryptic error message, but it's probably not appropriate to include in this PR which is only about the python version upgrade.

jlopp commented:
@bitcoinhodler please resolve the conflict by removing the import of random library

I'll also note that I tried running tests with that conflict resolved and got a failure:

```
cd testrun/create-deposit-data.test && ../../t/create-deposit-data.run 18333  2>&1 > ../../t/create-deposit-data.out
Traceback (most recent call last):
  File "../../glacierscript.py", line 858, in <module>
    deposit_interactive(args.m, args.n, args.dice, args.rng)
  File "../../glacierscript.py", line 649, in deposit_interactive
    addresses = [get_address_for_wif_privkey(key) for key in keys]
  File "../../glacierscript.py", line 649, in <listcomp>
    addresses = [get_address_for_wif_privkey(key) for key in keys]
  File "../../glacierscript.py", line 335, in get_address_for_wif_privkey
    label = hash_sha256(privkey)
  File "../../glacierscript.py", line 54, in hash_sha256
    m.update(s)
TypeError: Unicode-objects must be encoded before hashing
make: *** [Makefile:75: t/create-deposit-data.test] Error 1
```

bitcoinhodler commented:
What's the policy on rebasing? I could rebase my entire change onto master, and keep a clean history, which seems to be generally preferred among Git intelligentsia. But that will make review more difficult since (to be safe) you'll have to go over my entire change again.

Alternatively I could merge and patch up the two minor issues as part of the merge commit.

bitcoinhodler commented:
I went with the latter policy and merged in the latest master into this branch.

Alternatively, I created a [python3-rebased](https://github.com/bitcoinhodler/GlacierProtocol/tree/python3-rebased) branch that I could force-push over this one if we wanted to be cleaner about the history.

privacybuilder commented:
What is needed to move this PR along?

jlopp commented:
One more review would be nice if @jacoblyles @jazarija @gracenoah are available.

bitcoinhodler commented:
> 
> 
> Do we need to use encode here, given that hash_sha256 already does that?

I'm not sure how you left these comments on github, but it's not showing me what lines/changes you're referring to. My latest change on this branch (b75c400) took out a couple of unnecessary encodes. Does that address your concern?

bitcoinhodler commented:
> 
> 
> One more review would be nice if @jacoblyles @jazarija @gracenoah are available.

And while you're all here, how about a review on #76 and #73 that build atop this one.

I opened those before this one was merged so that people could review all of them at once, in an attempt to accelerate the pace of reviews.

I've got [6 more PRs](https://github.com/bitcoinhodler/GlacierProtocol/blob/master/README.md) ready for review after that, and then we get to PSBT, which I have working already.

privacybuilder commented:
@bitcoinhodler  - if you are looking for testers for the new combined PRs? Do you have them all merged in a single branch? Is there an update to the PDF that goes with those? How can I help? :D

bitcoinhodler commented:
> @bitcoinhodler - if you are looking for testers for the new combined PRs? Do you have them all merged in a single branch? Is there an update to the PDF that goes with those? How can I help? :D

I'm looking for reviewers, mainly. Read through my commits and make sure I haven't broken anything, or introduced any new weaknesses or backdoors (intentional or not).

Each of my three open PRs builds upon the previous (though they are out of order). Review this one first, then #76, then #73. The branch for #73 includes all the commits from all three PRs.

For these three PRs, and the next six I have ready to submit, no PDF updates are needed. It's not until we get to PSBT that the process changes at all, and that will no doubt be documented separately (as an optional appendix, perhaps) until it's gone through a public review process.

jacoblyles commented:
LGTM
